src/symbol/static_graph.cc

```diff
@@ -161,6 +161,97 @@ bool StaticGraph::InferNodeShapes(const std::vector<uint32_t> &topo_order,
   return true;
 }
 
+bool StaticGraph::InferNodeTypes(const std::vector<uint32_t> &topo_order,
```

Comment:
We should consider the node inference algorithm (type, shape, device) to a common pattern like visitor later